apps/invoices/src/modules/app-configuration/ui/address-form.tsx

```diff
@@ -1,79 +1,188 @@
+       *
+       * Alternative way will be fetching data in parent and set them in props (and dont render component if data is not set).
+       * This is caused by react hooks that are eager
+       *
+       * TODO it doesnt work with initial render (cmd+r)
+       */
+      reset();
+    }
+  }, [channelOverrideConfigQuery.data, reset, props.channelSlug]);
+
+  if (channelOverrideConfigQuery.isLoading) {
+    return <Text color={"textNeutralSubdued"}>Loading</Text>;
+  }
 
   return (
     <form
       onSubmit={handleSubmit((data, event) => {
-        props.onSubmit(data);
+        return upsertConfigMutation
+          .mutateAsync({
+            address: data,
+            channelSlug: props.channelSlug,
+          })
+          .then(() => {
+            notifySuccess("Success", "Updated channel configuration");
+
+            push("/configuration");
+          });
       })}
-      className={styles.form}
     >
-      <Typography component="h3" variant="h3" paragraph>
-        Configure
-        <span onClick={handleChannelNameClick} className={styles.channelName}>
-          {` ${props.channelName} `}
-        </span>
-        channel:
-      </Typography>
-      <TextField label="Company Name" {...CommonFieldProps} {...register("companyName")} />
-      <TextField label="Street Address 1" {...CommonFieldProps} {...register("streetAddress1")} />
-      <TextField {...CommonFieldProps} label="Street Address 2" {...register("streetAddress2")} />
-      <div style={{ display: "grid", gap: 20, gridTemplateColumns: "1fr 2fr" }}>
-        <TextField {...CommonFieldProps} label="Postal Code" {...register("postalCode")} />
-        <TextField {...CommonFieldProps} label="City" {...register("city")} />
-      </div>
-      <TextField {...CommonFieldProps} label="City Area" {...register("cityArea")} />
-      <TextField {...CommonFieldProps} label="Country" {...register("country")} />
-      <TextField label="Country Area" {...CommonFieldProps} {...register("countryArea")} />
-      <Button type="submit" fullWidth variant="primary">
-        Save channel configuration
-      </Button>
+      <Box display={"grid"} gap={6} marginBottom={12}>
+        {fieldsBlock1.map((fieldName) => (
+          <Controller
+            key={fieldName}
+            control={control}
+            render={({ field: { onChange, onBlur, value, name, ref } }) => {
+              return (
+                <Input
+                  onChange={onChange}
+                  value={value}
+                  error={!!formState.errors[fieldName]}
+                  label={fieldLabels[fieldName]}
+                  onBlur={onBlur}
+                  name={name}
+                  ref={ref}
+                />
+              );
+            }}
+            name={fieldName}
+          />
+        ))}
+
+        <Box display={"grid"} gridTemplateColumns={2} gap={6}>
+          {fieldsBlock2.map((fieldName) => (
```

Comment:
Yea, thats why forms suck :/ you are *almost* where you can abstract, but then, there is a one thing that breaks it

For now I think this is good-enough, when we have selects, obviously we need to rewrite. I don't know if we can ever find a good abstraction on forms (if we try, we end up having configurations longer than the template)

Reply:
As much as I would love to avoid bloat in the forms with some abstraction, there's always _something_ that abstraction did not take into account :(